### PR TITLE
feat: add option to set last active window as target for file actions

### PR DESCRIPTION
**Pull request by devxpain**
- Add new option `set_last_win_as_target` to `opts.actions`
- Update documentation to reflect new option
- Set default value of `set_last_win_as_target` to `true`

### CODE REVIEW

**Comment by alex-courtis** [comment]
Does `api.node.open.no_window_picker()` achieve the functionality you are after?

The `actions.open_file.window_picker` apply even when the `window_picker.enable` is false, as does the above API. This needs to be properly documented, see similar https://github.com/nvim-tree/nvim-tree.lua/discussions/3005

File: doc/nvim-tree-lua.txt [hunk 1]
```diff
@@ -570,6 +570,7 @@ Following is the default configuration. See |nvim-tree-opts| for details. >lua
       },
       actions = {
         use_system_clipboard = true,
+        set_last_win_as_target = true,
```
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:573]: This would be a change to existing behaviour; we'll need to default to false.

File: lua/nvim-tree.lua [hunk 2]
```diff
@@ -232,6 +232,17 @@ local function setup_autocommands(opts)
       end,
     })
   end
+
+  if opts.actions.set_last_win_as_target then
+    create_nvim_tree_autocmd("BufLeave", {
```
Comment (alex-courtis) [inline lua/nvim-tree.lua:237]: Please avoid autocommands; this should be done in `open-file.lua`

File: lua/nvim-tree.lua [hunk 3]
```diff
@@ -232,6 +232,17 @@ local function setup_autocommands(opts)
...
+      callback = function()
+        local buf_name = vim.api.nvim_buf_get_name(0)
+        if not string.find(buf_name, "NvimTree_") then
```
Comment (alex-courtis) [inline lua/nvim-tree.lua:240]: `utils.is_nvim_tree_buf` should be used.

File: lua/nvim-tree.lua [hunk 4]
```diff
@@ -232,6 +232,17 @@ local function setup_autocommands(opts)
       end,
     })
   end
+
+  if opts.actions.set_last_win_as_target then
+    create_nvim_tree_autocmd("BufLeave", {
+      callback = function()
+        local buf_name = vim.api.nvim_buf_get_name(0)
```
Comment (alex-courtis) [inline lua/nvim-tree.lua:239]: bufnr / bufname should come from the `ev` argument passed. See `:help nvim_create_autocmd()`

**Comment by alex-courtis** [comment]
Looking at similar issue #2262 `vim.fn.winnr("#")` may be a more reliable approach.

**Comment by devxpain** [comment]
I’ve updated the code as requested and reviewed the behavior of api.node.open.no_window_picker(). I didn’t encounter any issues.

Looking at similar issue #2262, I see vim.fn.winnr("#") mentioned as a more reliable approach. I’m not sure how it applies to this pull request. Could you clarify or provide an example?

Let me know if there’s anything else that needs adjustment!

**Comment by alex-courtis** [comment]
> Looking at similar issue #2262, I see vim.fn.winnr("#") mentioned as a more reliable approach. I’m not sure how it applies to this pull request. Could you clarify or provide an example?

The idea is to use the actual previous window, according to nvim, rather than the one tracked via `lib.target_winid`.

It could be something like:
```lua
local function get_target_winid(mode)
  local target_winid
  if not M.window_picker.enable or mode == "edit_no_picker" or mode == "preview_no_picker" then
    target_winnr = vim.fn.winnr("#")
```

We would likely need to check that against `usable_win_ids` and fall back to first.

Looking through other issues it seems that this is not the best approach: https://github.com/nvim-tree/nvim-tree.lua/issues/2262#issuecomment-1585963649

Let's stick with tracking.

File: lua/nvim-tree/actions/node/open-file.lua [hunk 5]
```diff
@@ -426,6 +426,15 @@ function M.setup(opts)
     opts.actions.open_file.window_picker.chars = tostring(opts.actions.open_file.window_picker.chars):upper()
   end
   M.window_picker = opts.actions.open_file.window_picker
+  if opts.actions.set_last_win_as_target then
+    vim.api.nvim_create_autocmd("BufLeave", {
+      callback = function()
+        if utils.is_nvim_tree_buf then
```
Comment (alex-courtis) [inline lua/nvim-tree/actions/node/open-file.lua:432]: This always evaluates to true, as it's not actually calling the function.

**Comment by alex-courtis** [comment]
I'm sorry, this is not a change we should make:
* window selection code is fragile enough already
* bypasses or invalidates other window selection options
* will have undesirable effects or possibly break in some cases when using multiple tabs; this will be resolved in https://github.com/nvim-tree/nvim-tree.lua/issues/2255

I think the best option for you is to write a `actions.open_file.window_picker.picker` function that exactly meets your use case.
You could perform your own window tracking and return your desired choice from that function.
You can even invoke a window picker in situations you choose e.g. `s1n7ax/nvim-window-picker`

**Pull request closed without merging**